Solution SOLID

fedaktaras opened the pull request:
Completed 
Using sealed classes

fedaktaras committed:
Task solved.
Using sealed classes.

fedaktaras committed:
Task solved.
Using sealed classes.

fedaktaras committed:
Task solved.
Using sealed classes.
It should have been using switch with sealed classes, but It has not passed test on java 17

TonyH277 commented on src/main/java/core/basesyntax/service/impl/ReaderCsv.java:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    static class DataProcessor {
+        private static final Set<String> TITLES_TO_REMOVE = Set.of("type,fruit,quantity");
+
+        public static List<String> filterTitles(List<String> data) {
+            return data.stream()
+                    .filter(row -> !TITLES_TO_REMOVE.contains(row))
+                    .collect(Collectors.toList());
+        }
+    }
```
could be replaced with lines.remove(0);

TonyH277 reviewed (changes requested):
Good job! Pay attention to constants

fedaktaras commented on src/main/java/core/basesyntax/model/Operation.java:
```diff
@@ -0,0 +1,26 @@
+package core.basesyntax.model;
+
+public abstract sealed class Operation permits Balance, Purchase, Supply, Return {
```
> Interesting aproach woth sealed classes. IMHO: Move all handlers from "model" to "strategy/impl" will be easier + if we had a lot of handlers -> you should permit them all (sound like extra work)

I will not be able to get invalid operation in any case. The best to use switch but it is in later java

fedaktaras commented on src/main/java/core/basesyntax/strategy/OperationStrategy.java:
```diff
@@ -0,0 +1,31 @@
+
+public class OperationStrategy {
+
+    public void processData(Operation operation) {
```
> It's more flexible if you use Map<> or switch/case for strategy

The idea was to use switch/case but with java 17 it doesn't work as required.

fedaktaras commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,26 @@
+        List<String> lines = reader.readData();
+        Parser parser = new ParserImpl();
+        List<Operation> parsedData = parser.parseToOperations(lines);
+        OperationStrategy operationStrategy = new OperationStrategy();
+        for (var pd : parsedData) {
+            operationStrategy.processData(pd);
+        }
```
> You could pass parsedData to operationStrategy trough constructor, and choose sertain strategy inside

It could be static field. It is created only once

fedaktaras committed:
Task solved.
Using sealed classes.
It should have been using switch with sealed classes, but It has not passed test on java 17

Changes after students review.

andrii-hoienko commented on pom.xml:
```diff
@@ -41,6 +41,14 @@
                     <includeTestSourceDirectory>true</includeTestSourceDirectory>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>17</source>
+                    <target>17</target>
+                </configuration>
+            </plugin>
```
What is the reason to add this?

fedaktaras replied:
> What is the reason to add this?

I made switch/case for sealed classes. I somehow changed project settings and it wasn`t pass tests on github.

andrii-hoienko commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,29 @@
+    private static final String READ_FILE_PATH = "src/files/fruits.csv";
+
+    public static void main(String[] args) {
+        Reader reader = new ReaderCsv(READ_FILE_PATH);
```
What if we want to reuse reader and read not only from this file READ_FILE_PATH? (same with writer)

andrii-hoienko commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,29 @@
+        Parser parser = new ParserImpl();
+        List<Operation> parsedData = parser.parseToOperations(lines);
+        OperationStrategy operationStrategy = new OperationStrategy();
+        for (var pd : parsedData) {
```
```suggestion
        for (var data: parsedData) {
```

andrii-hoienko commented on src/main/java/core/basesyntax/db/Storage.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.Map;
+
+public class Storage {
+    public static final Map<String, Integer> data = new HashMap<>();
```
Don`t forget about name conventions

andrii-hoienko commented on src/main/java/core/basesyntax/model/Purchase.java:
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.model;
+
+public final class Purchase extends Operation {
```
I suggest making Operation enum, operation objects always should be the same and in limited amount. And create FruitTransaction class to keep operation, amount and fruit name

andrii-hoienko commented on src/main/java/core/basesyntax/service/Parser.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.List;
+
+public interface Parser {
+    List<Operation> parseToOperations(List<String> record);
```
agree

andrii-hoienko commented on src/main/java/core/basesyntax/service/impl/ReaderCsv.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    static class DataProcessor {
+        private static final Set<String> TITLES_TO_REMOVE = Set.of("type,fruit,quantity");
```
What if we change the title, for example, we will add whitespace or something? In this case, we will go to the code and add a new value to the set. I suggest just skipping first element of the array to avoid the header

andrii-hoienko commented on src/main/java/core/basesyntax/service/impl/ReaderCsv.java:
```diff
@@ -0,0 +1,42 @@
+
+        public static List<String> filterTitles(List<String> data) {
+            return data.stream()
+                    .filter(row -> !TITLES_TO_REMOVE.contains(row))
```
```suggestion
                    .skip(1)
```

andrii-hoienko commented on src/main/java/core/basesyntax/service/impl/ReaderCsv.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    static class DataProcessor {
```
No reason to create separate nested class for that, just method is enough

andrii-hoienko commented on src/main/java/core/basesyntax/service/impl/WriterCsv.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public void writeData() {
+        try (FileWriter fileWriter = new FileWriter(filePath)) {
+            fileWriter.write("fruit,quantity" + System.lineSeparator());
+            for (Map.Entry<String, Integer> entry : Storage.data.entrySet()) {
+                fileWriter.write(entry.getKey() + "," + entry.getValue() + System.lineSeparator());
+            }
```
Please make string literals constants

andrii-hoienko commented on src/main/java/core/basesyntax/service/impl/WriterCsv.java:
```diff
@@ -0,0 +1,26 @@
+                fileWriter.write(entry.getKey() + "," + entry.getValue() + System.lineSeparator());
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Can`t write file.");
```
```suggestion
            throw new RuntimeException("Can`t write file:" + filePath);
```

andrii-hoienko reviewed (changes requested):
Good job, but some significant improvements required

fedaktaras committed:
Task solved.
Fixed all commented by mentor.

fedaktaras committed:
Task solved.
Fixed all commented by mentor.

yevheniisynytsia reviewed (approved):
Great job!
